core/core.go

```diff
@@ -222,62 +258,235 @@ func NewTextile(ctx context.Context, conf Config) (*Textile, error) {
 }
 
 func (t *Textile) Bootstrap() {
-	t.threadservice.Bootstrap(util.DefaultBoostrapPeers())
+	t.ts.Bootstrap(tutil.DefaultBoostrapPeers())
 }
 
 func (t *Textile) Close() error {
 	t.sessionBus.Discard()
 	if err := t.gateway.Stop(); err != nil {
 		return err
 	}
-	if err := t.threadsClient.Close(); err != nil {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	if err := t.proxy.Shutdown(ctx); err != nil {
 		return err
 	}
-	if err := t.threadservice.Close(); err != nil {
+	t.server.GracefulStop()
+	if err := t.th.Close(); err != nil {
 		return err
 	}
-	t.threadsServiceServer.Close()
-	t.threadsServer.Close()
-	if err := t.server.Close(); err != nil {
+	if err := t.ts.Close(); err != nil {
 		return err
 	}
-	return t.ds.Close()
+	if t.fc != nil {
+		if err := t.fc.Close(); err != nil {
+			return err
+		}
+	}
+	if err := t.co.Close(); err != nil {
+		return err
+	}
+	return nil
 }
 
 func (t *Textile) HostID() peer.ID {
-	return t.threadservice.Host().ID()
+	return t.ts.Host().ID()
 }
 
-func (t *Textile) clientAuthFunc(ctx context.Context) (context.Context, error) {
-	token, err := auth.AuthFromMD(ctx, "bearer")
-	if err != nil {
-		return nil, err
-	}
-	if token == t.threadsToken {
-		return ctx, nil
-	}
+func getService(method string) string {
+	return strings.Split(method, "/")[1]
+}
 
-	session, err := t.collections.Sessions.Get(ctx, token)
-	if err != nil {
-		return nil, status.Error(codes.Unauthenticated, "Invalid auth token")
+func (t *Textile) authFunc(ctx context.Context) (context.Context, error) {
+	method, _ := grpc.Method(ctx)
+	for _, ignored := range ignoreMethods {
+		if method == ignored {
+			return ctx, nil
+		}
 	}
-	if session.Expiry < int(time.Now().Unix()) {
-		return nil, status.Error(codes.Unauthenticated, "Expired auth token")
+
+	threadID, ok := common.ThreadIDFromMD(ctx)
+	if ok {
+		ctx = common.NewThreadIDContext(ctx, threadID)
```

Comment:
We want to pass the headers on to any additional requests, _which get picked up by the same interceptors_.